cmd/slsa-github-generator/attest.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2022 SLSA Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"regexp"
+	"strings"
+
+	intoto "github.com/in-toto/in-toto-golang/in_toto"
+	slsav02 "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v0.2"
+	"github.com/spf13/cobra"
+
+	"github.com/slsa-framework/slsa-github-generator/github"
+	"github.com/slsa-framework/slsa-github-generator/signing/sigstore"
+	"github.com/slsa-framework/slsa-github-generator/slsa"
+)
+
+var (
+	// shaCheck verifies a hash is has only hexidecimal digits and is 64
+	// characters long.
+	shaCheck = regexp.MustCompile(`^[a-fA-F0-9]{64}$`)
+)
+
+func parseSubjects(subjectsStr string) ([]intoto.Subject, error) {
+	var parsed []intoto.Subject
+
+	subjects := strings.Split(subjectsStr, "\n")
+	for _, s := range subjects {
+		parts := strings.SplitN(s, " ", 2)
+		if len(parts) == 0 {
+			return nil, errors.New("missing subject name")
+		}
+		// Do a sanity check on the SHA to make sure it's a proper hex digest.
+		shaDigest := strings.TrimSpace(parts[0])
+		if !shaCheck.MatchString(shaDigest) {
+			return nil, fmt.Errorf("unexpected sha256 hash %q", shaDigest)
+		}
+		if len(parts) == 1 {
+			return nil, fmt.Errorf("expected subject name for hash %q", shaDigest)
+		}
+		name := strings.TrimSpace(parts[1])
+		parsed = append(parsed, intoto.Subject{
+			Name: name,
+			Digest: slsav02.DigestSet{
+				// Lowercase the sha digest to comply with the SLSA spec.
+				"sha256": strings.ToLower(shaDigest),
+			},
+		})
+	}
+	return parsed, nil
```

Comment:
Check that each name is unique, as required by the spec. (OK to do in a follow-up PR.)

Reply:
#12